streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBStoreTest.java

```diff
@@ -1091,71 +1091,131 @@ public void shouldThrowWhenRestoringOnMissingHeaders() {
+            headers.add(new RecordHeader(
+                ChangelogRecordDeserializationHelper.VECTOR_KEY,
+                PositionSerde.serialize(position1).array()));
+            entries.add(new ConsumerRecord<>("", 0, 0L, RecordBatch.NO_TIMESTAMP,
+                TimestampType.NO_TIMESTAMP_TYPE, -1, -1,
                 "3".getBytes(UTF_8), "c".getBytes(UTF_8), headers, Optional.empty()));
+        }
         return entries;
     }
 
     private List<ConsumerRecord<byte[], byte[]>> getChangelogRecordsWithTombstones() {
         final List<ConsumerRecord<byte[], byte[]>> entries = new ArrayList<>();
-        final Headers headers = new RecordHeaders();
-        Position position = Position.emptyPosition();
-
-        position = position.update("A", 0, 1);
-        headers.add(ChangelogRecordDeserializationHelper.CHANGELOG_VERSION_HEADER_RECORD_CONSISTENCY);
-        headers.add(new RecordHeader(Position.VECTOR_KEY, position.serialize().array()));
-        entries.add(new ConsumerRecord<>("", 0, 0L,  RecordBatch.NO_TIMESTAMP, TimestampType.NO_TIMESTAMP_TYPE, -1, -1,
+        {
+            final Position position = Position
+                .emptyPosition()
+                .withComponent("A", 0, 1);
+            final Headers headers = new RecordHeaders();
+            headers.add(
+                ChangelogRecordDeserializationHelper.CHANGELOG_VERSION_HEADER_RECORD_CONSISTENCY);
+            headers.add(new RecordHeader(
+                ChangelogRecordDeserializationHelper.VECTOR_KEY,
+                PositionSerde.serialize(position).array()));
+            entries.add(new ConsumerRecord<>("", 0, 0L, RecordBatch.NO_TIMESTAMP,
+                TimestampType.NO_TIMESTAMP_TYPE, -1, -1,
                 "1".getBytes(UTF_8), "a".getBytes(UTF_8), headers, Optional.empty()));
-
-        position = position.update("A", 0, 2);
-        headers.add(ChangelogRecordDeserializationHelper.CHANGELOG_VERSION_HEADER_RECORD_CONSISTENCY);
-        headers.add(new RecordHeader(Position.VECTOR_KEY, position.serialize().array()));
-        entries.add(new ConsumerRecord<>("", 0, 0L,  RecordBatch.NO_TIMESTAMP, TimestampType.NO_TIMESTAMP_TYPE, -1, -1,
+        }
+        {
+            final Position position = Position
+                .emptyPosition()
+                .withComponent("A", 0, 2);
+            final Headers headers = new RecordHeaders();
+            headers.add(
+                ChangelogRecordDeserializationHelper.CHANGELOG_VERSION_HEADER_RECORD_CONSISTENCY);
+            headers.add(new RecordHeader(
+                ChangelogRecordDeserializationHelper.VECTOR_KEY,
+                PositionSerde.serialize(position).array()));
+            entries.add(new ConsumerRecord<>("", 0, 0L, RecordBatch.NO_TIMESTAMP,
+                TimestampType.NO_TIMESTAMP_TYPE, -1, -1,
                 "1".getBytes(UTF_8), null, headers, Optional.empty()));
-
+        }
```

Comment:
Most of the changes I made were just trivial conflict resolution, but you should be aware that there were several tests I discovered had this same bug. The Headers object was shared among all the records, so rather than sending several records with different headers, we were actually sending several objects with the same headers. To achieve the scenario you were going for, we need a fresh headers object each time.